stl/inc/algorithm

```diff
@@ -2348,6 +2348,83 @@ _FwdIt2 replace_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _
 }
 #endif // _HAS_CXX17
 
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // VARIABLE ranges::fill
+    class _Fill_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <class _Ty, output_iterator<const _Ty&> _It, sentinel_for<_It> _Se>
+        constexpr _It operator()(_It _First, _Se _Last, const _Ty& _Value) const {
+            _Adl_verify_range(_First, _Last);
+            auto _UFirst      = _Get_unwrapped(_STD move(_First));
+            const auto _ULast = _Get_unwrapped(_STD move(_Last));
+            if constexpr (_Fill_memset_is_safe<decltype(_UFirst), _Ty>) {
+#ifdef __cpp_lib_is_constant_evaluated
+                if (!_STD is_constant_evaluated())
+#endif // __cpp_lib_is_constant_evaluated
+                {
+                    _CSTD memset(_UFirst, static_cast<unsigned char>(_Value), static_cast<size_t>(_ULast - _UFirst));
+                    _UFirst += static_cast<size_t>(_ULast - _UFirst);
+                    _Seek_wrapped(_First, _UFirst);
+                    return _First;
+                }
+            }
+
+            while (_UFirst != _ULast) {
+                *_UFirst = _Value;
+                ++_UFirst;
+            }
+
+            _Seek_wrapped(_First, _STD move(_UFirst));
+            return _First;
+        }
+
+        template <class _Ty, output_range<const _Ty&> _Rng>
+        constexpr borrowed_iterator_t<_Rng> operator()(_Rng&& _Range, const _Ty& _Value) const {
+            return (*this)(_RANGES begin(_Range), _RANGES end(_Range), _Value);

```

Comment:
I am unsure of the conventions - should the range overloads pass `_RANGES _Ubegin` and `_RANGES _Uend` now, to avoid locking/unlocking when copying iterators? The `_Adl_verify_range` checking that the iterator overload performs is unnecessary when we're initially given a range (as we know we're correctly retrieving a begin and end into the same range, passed in the correct order).

Reply:
It would need to be a bit more complicated than simply `_Ubegin`/`_Uend` since we need to rewrap the returned iterator:
```suggestion
            auto _First = _RANGES begin(_Range);
            _Seek_wrapped(_First, (*this)(_Get_unwrapped(_STD move(_First)), _Uend(_Range), _Value));
            return _First;
```
which would be a reasonable change to make here (after, merging `master` to pull in the `_Ubegin`/`_Uend` changes). 

In the longer term, I think I'll be refactoring everything into a "public facades that call into private unchecked implementations" structure that will allow me to add Billy's "overloads instead of default arguments" overloads, possibly sharing "private unchecked implementations" with `std` algorithms using some fancy `if constexpr` footwork to get rid of `std::invoke` calls. [This exploratory reimplementation of `ranges::for_each`](https://github.com/CaseyCarter/STL/blob/afc249656130821320927ab5ec620ec16d0f6054/stl/inc/algorithm#L296-L364), for example, bloats up the maintained codesize a bit but generates code equivalent to `std::for_each` for "normal" function calls with `identity` projections.